File: Front-End/src/services/videoService.ts
import type { ApiResponse } from "@/types";
import { api } from "./api";

// ============================================================================
// LEGACY PATIENT VIDEO TYPES
// ============================================================================

export interface Video {
  id: string;
  patient_id: string;
  description?: string;
  file_path: string;
  public_video_url: string;
  created_at: string;
}

export interface VideoCreateRequest {
  patient_id: string;
  video_path: string;
  description?: string;
}

export interface VideoUpdateRequest {
  description?: string;
}

export interface VideoUploadResponse {
  id: string;
  patient_id: string;
  description?: string;
  file_path: string;
  public_video_url: string;
  created_at: string;
}

// ============================================================================
// AMBULANCE RECORDING TYPES (New)
// ============================================================================

export interface RecordingResponse {
  id: string;
  session_id: string;
  session_name?: string;
  ambulance_number?: string;
  camera_id?: string; // Camera identifier (e.g., "AMB-002-ROOM-001")
  recording_path?: string; // Local recording path
  file_path?: string;
  public_video_url?: string;
  duration?: number; // Duration in seconds
  file_size?: number; // File size in bytes
  session_start?: string;
  session_end?: string;
  created_at: string;
  is_archived: boolean; // True if uploaded to Supabase Storage
}

export interface SessionWithRecordings {
  session_id: string;
  session_name: string;
  ambulance_number: string;
  recordings: RecordingResponse[];
  total_duration: number;
  total_recordings: number;
}

// ============================================================================
// VIDEO SERVICE - LEGACY PATIENT VIDEOS
// ============================================================================

export const videoService = {
  // CRUD Operations for patient videos
  async getAll(): Promise<ApiResponse<Video[]>> {
    return api.get<Video[]>("/videos/");
  },

  async getByPatientId(patientId: string): Promise<ApiResponse<Video[]>> {
    return api.get<Video[]>(`/videos/${patientId}/videos`);
  },

  async create(data: VideoCreateRequest): Promise<ApiResponse<Video>> {
    return api.post<Video>("/videos/", data);
  },

  async delete(videoId: string): Promise<ApiResponse<{ message: string }>> {
    return api.delete<{ message: string }>(`/videos/${videoId}`);
  },
};

// ============================================================================
// RECORDING SERVICE - AMBULANCE SESSION RECORDINGS (New)
// ============================================================================

export const recordingService = {
  /**
   * Get all ambulance session recordings (both archived and live)
   */
  async getAllRecordings(): Promise<ApiResponse<RecordingResponse[]>> {
    return api.get<RecordingResponse[]>("/videos/recordings");
  },

  /**
   * Get only archived recordings (uploaded to Supabase Storage)
   * Excludes live/ongoing recordings
   */
  async getArchivedRecordings(): Promise<ApiResponse<RecordingResponse[]>> {
    return api.get<RecordingResponse[]>("/videos/recordings/archived");
  },

  /**
   * Get a specific recording by ID
   */
  async getRecordingById(
    recordingId: string
  ): Promise<ApiResponse<RecordingResponse>> {
    return api.get<RecordingResponse>(`/videos/recordings/${recordingId}`);
  },

  /**
   * Get all recordings for a specific ambulance session
   */
  async getRecordingsBySession(
    sessionId: string
  ): Promise<ApiResponse<RecordingResponse[]>> {
    return api.get<RecordingResponse[]>(
      `/videos/recordings/session/${sessionId}`
    );
  },

  /**
   * Get all recordings for a specific ambulance
   */
  async getRecordingsByAmbulance(
    ambulanceId: string
  ): Promise<ApiResponse<RecordingResponse[]>> {
    return api.get<RecordingResponse[]>(
      `/videos/recordings/ambulance/${ambulanceId}`
    );
  },

  /**
   * Delete a recording from the database
   * Note: This does NOT delete the file from Supabase Storage
   */
  async deleteRecording(
    recordingId: string
  ): Promise<ApiResponse<{ message: string }>> {
    return api.delete<{ message: string }>(`/videos/recordings/${recordingId}`);
  },

  /**
   * Group recordings by session
   * Helper function to organize recordings by session
   */
  async getSessionsWithRecordings(): Promise<
    ApiResponse<SessionWithRecordings[]>
  > {
    const response = await this.getArchivedRecordings();

    if (response.error || !response.data) {
      return {
        data: null,
        error: response.error || "Failed to fetch recordings",
      };
    }

    // Group recordings by session_id
    const sessionMap = new Map<string, RecordingResponse[]>();

    response.data.forEach((recording) => {
      const sessionId = recording.session_id;
      if (!sessionMap.has(sessionId)) {
        sessionMap.set(sessionId, []);
      }
      sessionMap.get(sessionId)!.push(recording);
    });

    // Convert to SessionWithRecordings array
    const sessions: SessionWithRecordings[] = Array.from(
      sessionMap.entries()
    ).map(([sessionId, recordings]) => {
      const totalDuration = recordings.reduce(
        (sum, rec) => sum + (rec.duration || 0),
        0
      );

      return {
        session_id: sessionId,
        session_name: recordings[0]?.session_name || "Unnamed Session",
        ambulance_number: recordings[0]?.ambulance_number || "Unknown",
        recordings,
        total_duration: totalDuration,
        total_recordings: recordings.length,
      };
    });

    // Sort by most recent first
    sessions.sort((a, b) => {
      const dateA = new Date(a.recordings[0]?.created_at || 0);
      const dateB = new Date(b.recordings[0]?.created_at || 0);
      return dateB.getTime() - dateA.getTime();
    });

    return {
      data: sessions,
      error: null,
    };
  },
};

File: Back-End/services/video_service.py
"""
Video service for managing ambulance session recordings from Supabase Storage.
Handles retrieval of recorded videos for playback in the frontend.
"""

from typing import Optional, List, Dict, Any
from datetime import datetime
from core.common import supabase, logger


class VideoService:
    """Service for handling video recording operations."""

    @staticmethod
    async def get_all_recordings() -> List[Dict[str, Any]]:
        """
        Get all ambulance session recordings with public URLs.
        Returns videos from Supabase Storage.
        """
        try:
            result = (
                supabase.table("ambulance_session_recordings")
                .select(
                    """
                    *,
                    ambulance_streaming_sessions!inner(
                        session_name,
                        ambulance_id,
                        ambulances(ambulance_number, license_plate)
                    )
                """
                )
                .order("created_at", desc=True)
                .execute()
            )

            recordings = []
            for record in result.data or []:
                # Get public URL from Supabase Storage if available
                public_url = None
                if record.get("storage_url"):
                    public_url = record["storage_url"]
                elif record.get("hls_playlist_url"):
                    # Fallback to HLS URL for live sessions
                    public_url = record["hls_playlist_url"]

                recordings.append(
                    {
                        "id": record["id"],
                        "session_id": record["session_id"],
                        "session_name": record.get(
                            "ambulance_streaming_sessions", {}
                        ).get("session_name"),
                        "ambulance_number": record.get(
                            "ambulance_streaming_sessions", {}
                        )
                        .get("ambulances", {})
                        .get("ambulance_number"),
                        "file_path": record.get("storage_url")
                        or record.get("hls_playlist_url"),
                        "public_video_url": public_url,
                        "duration": record.get("duration"),
                        "file_size": record.get("file_size"),
                        "session_start": record.get("session_start"),
                        "session_end": record.get("session_end"),
                        "created_at": record["created_at"],
                        "is_archived": record.get("storage_url")
                        is not None,  # True if uploaded to Supabase
                    }
                )

            return recordings

        except Exception as e:
            logger.error("Error fetching all recordings: %s", e)
            raise

    @staticmethod
    async def get_recording_by_id(recording_id: str) -> Optional[Dict[str, Any]]:
        """
        Get a specific recording by ID with public URL.
        """
        try:
            result = (
                supabase.table("ambulance_session_recordings")
                .select(
                    """
                    *,
                    ambulance_streaming_sessions!inner(
                        session_name,
                        ambulance_id,
                        ambulances(ambulance_number, license_plate)
                    )
                """
                )
                .eq("id", recording_id)
                .single()
                .execute()
            )

            if not result.data:
                return None

            record = result.data

            # Get public URL from Supabase Storage if available
            public_url = None
            if record.get("storage_url"):
                public_url = record["storage_url"]
            elif record.get("hls_playlist_url"):
                # Fallback to HLS URL for live sessions
                public_url = record["hls_playlist_url"]

            return {
                "id": record["id"],
                "session_id": record["session_id"],
                "session_name": record.get("ambulance_streaming_sessions", {}).get(
                    "session_name"
                ),
                "ambulance_number": record.get("ambulance_streaming_sessions", {})
                .get("ambulances", {})
                .get("ambulance_number"),
                "file_path": record.get("storage_url")
                or record.get("hls_playlist_url"),
                "public_video_url": public_url,
                "duration": record.get("duration"),
                "file_size": record.get("file_size"),
                "session_start": record.get("session_start"),
                "session_end": record.get("session_end"),
                "created_at": record["created_at"],
                "is_archived": record.get("storage_url")
                is not None,  # True if uploaded to Supabase
            }

        except Exception as e:
            logger.error("Error fetching recording %s: %s", recording_id, e)
            raise

    @staticmethod
    async def get_recordings_by_session(session_id: str) -> List[Dict[str, Any]]:
        """
        Get all recordings for a specific ambulance session.
        """
        try:
            result = (
                supabase.table("ambulance_session_recordings")
                .select("*")
                .eq("session_id", session_id)
                .order("created_at", desc=True)
                .execute()
            )

            recordings = []
            for record in result.data or []:
                recordings.append(
                    {
                        "id": record["id"],
                        "session_id": record["session_id"],
                        "session_name": None,
                        "ambulance_number": None,
                        "camera_id": record.get("camera_id"),
                        "recording_path": record.get("recording_path"),
                        "file_path": record.get("storage_url"),
                        "public_video_url": record.get("storage_url"),
                        "duration": record.get("duration"),
                        "file_size": record.get("file_size"),
                        "session_start": record.get("session_start"),
                        "session_end": record.get("session_end"),
                        "created_at": record["created_at"],
                        "is_archived": record.get("storage_url") is not None,
                    }
                )

            return recordings

        except Exception as e:
            logger.error("Error fetching recordings for session %s: %s", session_id, e)
            raise

    @staticmethod
    async def get_recordings_by_ambulance(ambulance_id: str) -> List[Dict[str, Any]]:
        """
        Get all recordings for a specific ambulance.
        """
        try:
            result = (
                supabase.table("ambulance_session_recordings")
                .select(
                    """
                    *,
                    ambulance_streaming_sessions!inner(
                        session_name,
                        ambulance_id,
                        ambulances(ambulance_number, license_plate)
                    )
                """
                )
                .eq("ambulance_streaming_sessions.ambulance_id", ambulance_id)
                .order("created_at", desc=True)
                .execute()
            )

            recordings = []
            for record in result.data or []:
                # Get public URL from Supabase Storage if available
                public_url = None
                if record.get("storage_url"):
                    public_url = record["storage_url"]
                elif record.get("hls_playlist_url"):
                    # Fallback to HLS URL for live sessions
                    public_url = record["hls_playlist_url"]

                recordings.append(
                    {
                        "id": record["id"],
                        "session_id": record["session_id"],
                        "session_name": record.get(
                            "ambulance_streaming_sessions", {}
                        ).get("session_name"),
                        "ambulance_number": record.get(
                            "ambulance_streaming_sessions", {}
                        )
                        .get("ambulances", {})
                        .get("ambulance_number"),
                        "file_path": record.get("storage_url")
                        or record.get("hls_playlist_url"),
                        "public_video_url": public_url,
                        "duration": record.get("duration"),
                        "file_size": record.get("file_size"),
                        "session_start": record.get("session_start"),
                        "session_end": record.get("session_end"),
                        "created_at": record["created_at"],
                        "is_archived": record.get("storage_url")
                        is not None,  # True if uploaded to Supabase
                    }
                )

            return recordings

        except Exception as e:
            logger.error(
                "Error fetching recordings for ambulance %s: %s", ambulance_id, e
            )
            raise

    @staticmethod
    async def get_archived_recordings() -> List[Dict[str, Any]]:
        """
        Get only archived recordings (uploaded to Supabase Storage).
        Excludes live/ongoing recordings that only have HLS URLs.
        """
        try:
            result = (
                supabase.table("ambulance_session_recordings")
                .select(
                    """
                    *,
                    ambulance_streaming_sessions!inner(
                        session_name,
                        ambulance_id,
                        ambulances(ambulance_number, license_plate)
                    )
                """
                )
                .not_.is_("storage_url", "null")  # Only records with storage_url
                .order("created_at", desc=True)
                .execute()
            )

            recordings = []
            for record in result.data or []:
                recordings.append(
                    {
                        "id": record["id"],
                        "session_id": record["session_id"],
                        "session_name": record.get(
                            "ambulance_streaming_sessions", {}
                        ).get("session_name"),
                        "ambulance_number": record.get(
                            "ambulance_streaming_sessions", {}
                        )
                        .get("ambulances", {})
                        .get("ambulance_number"),
                        "file_path": record.get("storage_url"),
                        "public_video_url": record.get("storage_url"),
                        "duration": record.get("duration"),
                        "file_size": record.get("file_size"),
                        "session_start": record.get("session_start"),
                        "session_end": record.get("session_end"),
                        "created_at": record["created_at"],
                        "is_archived": True,
                    }
                )

            return recordings

        except Exception as e:
            logger.error("Error fetching archived recordings: %s", e)
            raise

    @staticmethod
    async def delete_recording(recording_id: str) -> bool:
        """
        Delete a recording from the database.
        Note: This does NOT delete the file from Supabase Storage.
        For storage deletion, use Supabase Storage admin operations.
        """
        try:
            result = (
                supabase.table("ambulance_session_recordings")
                .delete()
                .eq("id", recording_id)
                .execute()
            )

            return True

        except Exception as e:
            logger.error("Error deleting recording %s: %s", recording_id, e)
            raise
